crates/ruma-macros/src/events/event_enum.rs

```diff
@@ -156,22 +162,36 @@ fn expand_event_enum(
 fn expand_deserialize_impl(
     kind: EventKind,
     var: EventEnumVariation,
-    events: &[(LitStr, Path)],
-    variants: &[EventEnumVariant],
+    events: &[EventEnumEntry],
     ruma_common: &TokenStream,
 ) -> syn::Result<TokenStream> {
     let serde = quote! { #ruma_common::exports::serde };
     let serde_json = quote! { #ruma_common::exports::serde_json };
 
     let ident = kind.to_event_enum_ident(var.into())?;
 
-    let variant_attrs = variants.iter().map(|v| {
-        let attrs = &v.attrs;
-        quote! { #(#attrs)* }
-    });
-    let self_variants = variants.iter().map(|v| v.ctor(quote! { Self }));
-    let content = events.iter().map(|(name, path)| to_event_path(name, path, kind, var));
-    let event_types: Vec<_> = events.iter().map(|(name, _)| name).collect();
+    let mut match_arms = Vec::with_capacity(events.len());
+    for event in events.iter() {
+        let variant = event.to_variant()?;
+        let variant_attrs = {
+            let attrs = &variant.attrs;
+            quote! { #(#attrs)* }
+        };
+        let self_variant = variant.ctor(quote! { Self });
+        let content = to_event_path(event.stable_name()?, &event.ev_path, kind, var);
+        let mut ev_types = event.aliases.clone();
+        ev_types.push(event.ev_type.clone());
+
+        for ev_type in ev_types {
+            match_arms.push(quote! {
+                #variant_attrs #ev_type => {
+                    let event = #serde_json::from_str::<#content>(json.get())
+                        .map_err(D::Error::custom)?;
+                    Ok(#self_variant(event))
+                },
+            });
+        }
```

Comment:
Also I think instead of the loop w/ `push` it should be possible to do this as `let match_arms: TokenStream = events.iter().map(/* ... */).collect::<syn::Result<_>>()?;`.